{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import sys\n",
    "import argparse\n",
    "import logging\n",
    "import subprocess\n",
    "from tabulate import tabulate\n",
    "from distutils.version import StrictVersion\n",
    "\n",
    "from dbclients.tantalus import TantalusApi\n",
    "from dbclients.colossus import ColossusApi\n",
    "from dbclients.basicclient import NotFoundError\n",
    "\n",
    "tantalus_api = TantalusApi()\n",
    "colossus_api = ColossusApi()\n",
    "\n",
    "JIRA_TICKET_REGEX = r\".*SC-\\d{4}$\"\n",
    "\n",
    "def find_latest_analysis_ticket(pool_id):\n",
    "    '''\n",
    "    Return the jira ticket of the newest possible analysis of the given library.\n",
    "    '''\n",
    "    analyses = list(colossus_api.list(\"analysis_information\", library__pool_id = pool_id))\n",
    "    latest_completion_version = None\n",
    "    analysis_jira_ticket = None\n",
    "    newest_version = False\n",
    "    #loop over the list of analyses to find the latest possible one\n",
    "    for analysis in analyses:\n",
    "        status = analysis[\"analysis_run\"][\"run_status\"]\n",
    "        # print(status)\n",
    "        current_version = analysis[\"version\"][1:]\n",
    "        # print(current_version)\n",
    "        if status in [\"complete\", \"hmmcopy_complete\"]:\n",
    "            if latest_completion_version:\n",
    "                if StrictVersion(current_version.strip('v')) >= StrictVersion(latest_completion_version.strip('v')):\n",
    "                    #print(\"The updated version %s\" % (current_version))\n",
    "                    latest_completion_version = current_version\n",
    "                    analysis_jira_ticket = analysis[\"analysis_jira_ticket\"]\n",
    "                    current_analysis = analysis\n",
    "            else:\n",
    "                latest_completion_version = current_version\n",
    "                analysis_jira_ticket = analysis[\"analysis_jira_ticket\"]\n",
    "                current_analysis = analysis\n",
    "                newest_version = True\n",
    "            #print(status, current_version, latest_completion_version, current_version > latest_completion_version)\n",
    "        else:\n",
    "            pass\n",
    "            #if the latest COMPLETE analysis exisis, save the information to the dict.\n",
    "\n",
    "    if newest_version:\n",
    "        return analysis_jira_ticket\n",
    "\n",
    "    return None\n",
    "\n",
    "def check_user_input(intial_question, options):\n",
    "\n",
    "    answer = input(intial_question)\n",
    "\n",
    "    while answer not in options:\n",
    "        answer = input(\"Invalid input. Options are {}: \".format(options))\n",
    "\n",
    "        if answer in options:\n",
    "            break\n",
    "\n",
    "    return answer\n",
    "\n",
    "\n",
    "def check_user_file_input(user_input):\n",
    "    file_choices = [\"metrics\", \"reads\", \"segments\", \"plots\"]\n",
    "    if user_input.lower() == \"y\":\n",
    "        return \"all\"\n",
    "\n",
    "    elif user_input.lower() == \"n\":\n",
    "        file_table = tabulate([[f] for f in file_choices], showindex=\"always\")\n",
    "        file_question = \"\\n{}\\n\\nPlease enter the index of the desired file: \".format(file_table)\n",
    "\n",
    "        file_index = check_user_input(file_question, list(map(str, range(len(file_choices)))))\n",
    "\n",
    "    print(\"\\nPreparing to download {} files...\\n\".format(file_choices[int(file_index)]))\n",
    "    return file_choices[int(file_index)]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def get_segment(jira_ticket, version, library_id):\n",
    "\n",
    "#     filenames = []\n",
    "#     file_resources = list(tantalus_api.list(\n",
    "#             \"file_resource\",\n",
    "#             filename__startswith=jira_ticket,\n",
    "#             filename__endswith=\"{}_segments.csv.gz\".format(library_id)\n",
    "#         ))\n",
    "\n",
    "#     filenames += [f[\"filename\"] for f in file_resources]\n",
    "\n",
    "#     return filenames\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_haplotype(jira_ticket, version, library_id):\n",
    "\n",
    "    filenames = []\n",
    "    file_resources = list(tantalus_api.list(\n",
    "            \"file_resource\",\n",
    "            filename__startswith=jira_ticket,\n",
    "            filename__endswith=\"allele_counts.tsv\"\n",
    "        ))\n",
    "\n",
    "    filenames += [f[\"filename\"] for f in file_resources]\n",
    "\n",
    "    return filenames\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_all_files(jira_ticket, version, library_id):\n",
    "    filenames = []\n",
    "\n",
    "    filenames += get_haplotype(jira_ticket, version, library_id)\n",
    "    return filenames\n",
    "\n",
    "def main(library_id, download_dir):\n",
    "    jira_ticket = find_latest_analysis_ticket(library_id) \n",
    "    # Hoa, add condition\n",
    "    if jira_ticket:\n",
    "        print('Latest Jira ticket  is: {0} correspond to library id: {1}'.format(jira_ticket,library_id))\n",
    "        # Define source and destination storage clients\n",
    "        from_storage_name = \"singlecellresults\"\n",
    "        to_storage_name = download_dir\n",
    "\n",
    "        from_storage = tantalus_api.get_storage(from_storage_name)\n",
    "        from_storage_client = tantalus_api.get_storage_client(from_storage_name)\n",
    "\n",
    "        # Check if user input is a jira ticket\n",
    "        # If so, check if its a analysis ticket.\n",
    "        # Else check if its a library ticket and return analysis tickets\n",
    "\n",
    "        if re.match(JIRA_TICKET_REGEX, jira_ticket):\n",
    "\n",
    "            try:\n",
    "                analysis_object = colossus_api.get(\"analysis_information\", analysis_jira_ticket=jira_ticket)\n",
    "                analysis_ticket = jira_ticket\n",
    "                # library_id = analysis_object[\"library\"][\"pool_id\"]\n",
    "                print(\"{} is a valid analysis ticket\\n\".format(analysis_ticket))\n",
    "\n",
    "            except NotFoundError:\n",
    "                raise Exception(\"{} is not an analysis ticket\".format(jira_ticket))\n",
    "\n",
    "        version = analysis_object[\"version\"]\n",
    "        filenames = get_all_files(analysis_ticket, version, library_id)\n",
    "        downloaded_files = []\n",
    "        for filename in filenames:\n",
    "            filepath_parsed = filename.split(\"/\")\n",
    "            cell_state = filepath_parsed[1]\n",
    "            analysis_type = filepath_parsed[-2]\n",
    "            file = filepath_parsed[-1]\n",
    "            print(\"cell state {0}; analysis type: {1};  files: {2}\".format(cell_state,analysis_type,file))\n",
    "\n",
    "            # REFACTOR\n",
    "            # This takes care of results from old version of scpipeline\n",
    "            if analysis_type == \"plots\":\n",
    "                if \"heatmap\" in filename:\n",
    "                    analysis_type = \"hmmcopy_autoploidy\"\n",
    "                else:\n",
    "                    analysis_type = \"alignment\"\n",
    "\n",
    "            # accommadates weird cell_state_prediction naming /storage convention\n",
    "            if cell_state == \"cell_state_prediction\":\n",
    "                subdir = os.path.join(to_storage_name, library_id, cell_state)\n",
    "                filepath = os.path.join(to_storage_name, library_id, cell_state, \"{}_cell_state_prediction.csv\".format(library_id))\n",
    "            else:\n",
    "                subdir = os.path.join(to_storage_name, library_id, analysis_type)\n",
    "                filepath = os.path.join(to_storage_name, library_id, analysis_type, file)\n",
    "\n",
    "            if not os.path.exists(subdir):\n",
    "                os.makedirs(subdir)\n",
    "\n",
    "            print(\"Downloading {} to {}\".format(file, subdir))\n",
    "\n",
    "            try:\n",
    "                blob = from_storage_client.blob_service.get_blob_to_path(\n",
    "                    container_name=\"results\",\n",
    "                    blob_name=filename,\n",
    "                    file_path=filepath\n",
    "                )\n",
    "                downloaded_files.append(filepath)\n",
    "            except:\n",
    "                print('{} does not exist in blob.'.format(filename.split(\"/\")[-1]))\n",
    "                pass\n",
    "\n",
    "        #clean up dir and unzip\n",
    "        for root, dirs, files in os.walk(os.path.join(to_storage_name, library_id)):\n",
    "            for name in files:\n",
    "                if (os.path.getsize(os.path.join(root, name)) == 0 and re.search(\".gz$\", os.path.join(root, name))):\n",
    "                    rm_cmd = [\"rm\", \"-f\", os.path.join(root, name)]\n",
    "                    subprocess.check_call(rm_cmd)\n",
    "                    continue\n",
    "\n",
    "                elif re.search(\".gz$\", os.path.join(root, name)):\n",
    "                    gunzip_cmd = [\"gunzip\", \"-f\", os.path.join(root, name)]\n",
    "                    subprocess.check_call(gunzip_cmd)\n",
    "\n",
    "\n",
    "        print(\"\\n********** Download complete **********\\n\\n\")\n",
    "        for file in downloaded_files:\n",
    "            print(file)\n",
    "            \n",
    "    else:\n",
    "        print(\"Jira ticket do not exist for library id: {0}. Please double check the analysis\".format(library_id))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(16, 6)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['Sample_ID', 'PDX_ID', 'DLP+_jira_ticket', 'Library ID_DLP+',\n",
       "       'Results_DLP', 'DLP_analysis'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "download_dir = '/home/htran/storage/datasets/hakwoo_metastasis'\n",
    "import pandas as pd\n",
    "meta_sample = pd.read_csv(os.path.join(download_dir,'SA919X7_metadata_Hoa.csv'),header=0)\n",
    "# meta_sample = meta_sample.loc[i for i in meta_sample.index if meta_sample['DLP_analysis']=='TRUE',:]\n",
    "\n",
    "print(meta_sample.shape)\n",
    "# ls_rows = [i for i in meta_sample.index if meta_sample.loc[i,'DLP_analysis']=='TRUE']\n",
    "# print(ls_rows)\n",
    "# len(library_ids)\n",
    "# meta_sample.values\n",
    "meta_sample.columns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(14, 6)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_sample = meta_sample[meta_sample['DLP_analysis']==True]\n",
    "meta_sample.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Latest Jira ticket  is: SC-3755 correspond to library id: A96200A\n",
      "SC-3755 is a valid analysis ticket\n",
      "\n",
      "cell state results; analysis type: sample_SA919X7XB05402;  files: allele_counts.tsv\n",
      "Downloading allele_counts.tsv to /home/htran/storage/datasets/hakwoo_metastasis/A96200A/sample_SA919X7XB05402\n",
      "\n",
      "********** Download complete **********\n",
      "\n",
      "\n",
      "/home/htran/storage/datasets/hakwoo_metastasis/A96200A/sample_SA919X7XB05402/allele_counts.tsv\n",
      "Latest Jira ticket  is: SC-3756 correspond to library id: A96204B\n",
      "SC-3756 is a valid analysis ticket\n",
      "\n",
      "cell state results; analysis type: sample_SA919X7XB05691;  files: allele_counts.tsv\n",
      "Downloading allele_counts.tsv to /home/htran/storage/datasets/hakwoo_metastasis/A96204B/sample_SA919X7XB05691\n",
      "\n",
      "********** Download complete **********\n",
      "\n",
      "\n",
      "/home/htran/storage/datasets/hakwoo_metastasis/A96204B/sample_SA919X7XB05691/allele_counts.tsv\n",
      "Latest Jira ticket  is: SC-3137 correspond to library id: A96229A\n",
      "SC-3137 is a valid analysis ticket\n",
      "\n",
      "cell state results; analysis type: sample_SA919X7XB05692;  files: allele_counts.tsv\n",
      "Downloading allele_counts.tsv to /home/htran/storage/datasets/hakwoo_metastasis/A96229A/sample_SA919X7XB05692\n",
      "\n",
      "********** Download complete **********\n",
      "\n",
      "\n",
      "/home/htran/storage/datasets/hakwoo_metastasis/A96229A/sample_SA919X7XB05692/allele_counts.tsv\n",
      "Latest Jira ticket  is: SC-2911 correspond to library id: A96117B\n",
      "SC-2911 is a valid analysis ticket\n",
      "\n",
      "cell state results; analysis type: sample_SA919X7XB05378;  files: allele_counts.tsv\n",
      "Downloading allele_counts.tsv to /home/htran/storage/datasets/hakwoo_metastasis/A96117B/sample_SA919X7XB05378\n",
      "\n",
      "********** Download complete **********\n",
      "\n",
      "\n",
      "/home/htran/storage/datasets/hakwoo_metastasis/A96117B/sample_SA919X7XB05378/allele_counts.tsv\n",
      "Latest Jira ticket  is: SC-3748 correspond to library id: A96117A\n",
      "SC-3748 is a valid analysis ticket\n",
      "\n",
      "cell state results; analysis type: sample_SA919X7XB05604;  files: allele_counts.tsv\n",
      "Downloading allele_counts.tsv to /home/htran/storage/datasets/hakwoo_metastasis/A96117A/sample_SA919X7XB05604\n",
      "\n",
      "********** Download complete **********\n",
      "\n",
      "\n",
      "/home/htran/storage/datasets/hakwoo_metastasis/A96117A/sample_SA919X7XB05604/allele_counts.tsv\n",
      "Latest Jira ticket  is: SC-3268 correspond to library id: A98300B\n",
      "SC-3268 is a valid analysis ticket\n",
      "\n",
      "cell state results; analysis type: sample_SA919X7XB05298;  files: allele_counts.tsv\n",
      "Downloading allele_counts.tsv to /home/htran/storage/datasets/hakwoo_metastasis/A98300B/sample_SA919X7XB05298\n",
      "\n",
      "********** Download complete **********\n",
      "\n",
      "\n",
      "/home/htran/storage/datasets/hakwoo_metastasis/A98300B/sample_SA919X7XB05298/allele_counts.tsv\n",
      "Latest Jira ticket  is: SC-3273 correspond to library id: A98290A\n",
      "SC-3273 is a valid analysis ticket\n",
      "\n",
      "cell state results; analysis type: sample_SA919X7XB05588;  files: allele_counts.tsv\n",
      "Downloading allele_counts.tsv to /home/htran/storage/datasets/hakwoo_metastasis/A98290A/sample_SA919X7XB05588\n",
      "\n",
      "********** Download complete **********\n",
      "\n",
      "\n",
      "/home/htran/storage/datasets/hakwoo_metastasis/A98290A/sample_SA919X7XB05588/allele_counts.tsv\n",
      "Latest Jira ticket  is: SC-3110 correspond to library id: A98299A\n",
      "SC-3110 is a valid analysis ticket\n",
      "\n",
      "cell state results; analysis type: sample_SA919X7XB05588;  files: allele_counts.tsv\n",
      "Downloading allele_counts.tsv to /home/htran/storage/datasets/hakwoo_metastasis/A98299A/sample_SA919X7XB05588\n",
      "\n",
      "********** Download complete **********\n",
      "\n",
      "\n",
      "/home/htran/storage/datasets/hakwoo_metastasis/A98299A/sample_SA919X7XB05588/allele_counts.tsv\n",
      "Latest Jira ticket  is: SC-3097 correspond to library id: A98293A\n",
      "SC-3097 is a valid analysis ticket\n",
      "\n",
      "cell state results; analysis type: sample_SA919X7XB05608;  files: allele_counts.tsv\n",
      "Downloading allele_counts.tsv to /home/htran/storage/datasets/hakwoo_metastasis/A98293A/sample_SA919X7XB05608\n",
      "\n",
      "********** Download complete **********\n",
      "\n",
      "\n",
      "/home/htran/storage/datasets/hakwoo_metastasis/A98293A/sample_SA919X7XB05608/allele_counts.tsv\n",
      "Latest Jira ticket  is: SC-3178 correspond to library id: A98232B\n",
      "SC-3178 is a valid analysis ticket\n",
      "\n",
      "cell state results; analysis type: sample_SA919X3XB08939;  files: allele_counts.tsv\n",
      "Downloading allele_counts.tsv to /home/htran/storage/datasets/hakwoo_metastasis/A98232B/sample_SA919X3XB08939\n",
      "\n",
      "********** Download complete **********\n",
      "\n",
      "\n",
      "/home/htran/storage/datasets/hakwoo_metastasis/A98232B/sample_SA919X3XB08939/allele_counts.tsv\n",
      "Latest Jira ticket  is: SC-3188 correspond to library id: A98306A\n",
      "SC-3188 is a valid analysis ticket\n",
      "\n",
      "cell state results; analysis type: sample_SA919X3XB08939;  files: allele_counts.tsv\n",
      "Downloading allele_counts.tsv to /home/htran/storage/datasets/hakwoo_metastasis/A98306A/sample_SA919X3XB08939\n",
      "\n",
      "********** Download complete **********\n",
      "\n",
      "\n",
      "/home/htran/storage/datasets/hakwoo_metastasis/A98306A/sample_SA919X3XB08939/allele_counts.tsv\n",
      "Latest Jira ticket  is: SC-3212 correspond to library id: A98261B\n",
      "SC-3212 is a valid analysis ticket\n",
      "\n",
      "cell state results; analysis type: sample_SA919X4XB09563;  files: allele_counts.tsv\n",
      "Downloading allele_counts.tsv to /home/htran/storage/datasets/hakwoo_metastasis/A98261B/sample_SA919X4XB09563\n",
      "\n",
      "********** Download complete **********\n",
      "\n",
      "\n",
      "/home/htran/storage/datasets/hakwoo_metastasis/A98261B/sample_SA919X4XB09563/allele_counts.tsv\n",
      "Latest Jira ticket  is: SC-3348 correspond to library id: A98217B\n",
      "SC-3348 is a valid analysis ticket\n",
      "\n",
      "cell state results; analysis type: sample_SA919X4XB40503;  files: allele_counts.tsv\n",
      "Downloading allele_counts.tsv to /home/htran/storage/datasets/hakwoo_metastasis/A98217B/sample_SA919X4XB40503\n",
      "\n",
      "********** Download complete **********\n",
      "\n",
      "\n",
      "/home/htran/storage/datasets/hakwoo_metastasis/A98217B/sample_SA919X4XB40503/allele_counts.tsv\n",
      "Latest Jira ticket  is: SC-3309 correspond to library id: A98217A\n",
      "SC-3309 is a valid analysis ticket\n",
      "\n",
      "cell state results; analysis type: sample_SA919X4XB40509;  files: allele_counts.tsv\n",
      "Downloading allele_counts.tsv to /home/htran/storage/datasets/hakwoo_metastasis/A98217A/sample_SA919X4XB40509\n",
      "\n",
      "********** Download complete **********\n",
      "\n",
      "\n",
      "/home/htran/storage/datasets/hakwoo_metastasis/A98217A/sample_SA919X4XB40509/allele_counts.tsv\n"
     ]
    }
   ],
   "source": [
    "library_ids = meta_sample['Library ID_DLP+']\n",
    "for library_id in library_ids:\n",
    "    main(library_id, download_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
